plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -96,19 +96,44 @@ public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path p
             return fs.listLocatedStatus(path);
         }
 
-        ValueHolder cachedValueHolder = uncheckedCacheGet(cache, path, ValueHolder::new);
+        return listInternal(fs, new DirectoryListingCacheKey(path, false));
+    }
+
+    @Override
+    public RemoteIterator<LocatedFileStatus> listFilesRecursively(FileSystem fs, Table table, Path path)
+            throws IOException
+    {
+        if (!isCacheEnabledFor(table.getSchemaTableName())) {
+            return fs.listFiles(path, true);
+        }
+
+        return listInternal(fs, new DirectoryListingCacheKey(path, true));
+    }
+
+    private RemoteIterator<LocatedFileStatus> listInternal(FileSystem fs, DirectoryListingCacheKey cacheKey)
+            throws IOException
+    {
+        ValueHolder cachedValueHolder = uncheckedCacheGet(cache, cacheKey, ValueHolder::new);
         if (cachedValueHolder.getFiles().isPresent()) {
             return new SimpleRemoteIterator(cachedValueHolder.getFiles().get().iterator());
         }
-        return cachingRemoteIterator(cachedValueHolder, fs.listLocatedStatus(path), path);
+
+        RemoteIterator<LocatedFileStatus> iterator;
+        if (cacheKey.isRecursiveFilesOnly()) {
+            iterator = fs.listFiles(cacheKey.getPath(), true);
+        }
+        else {
+            iterator = fs.listLocatedStatus(cacheKey.getPath());
```

Comment:
Why not always call `fs.listFiles(cachKey.getPath(), cacheKey.recursiveFilesOnly()`?

Reply:
`fs.listFiles(Path, boolean recursive)` only lists files, not directories (and does so either recursively, or only at the current level), while `fs.listLocatedStatus(Path)` returns both.